client_mining_p/miner.py

```diff
@@ -17,9 +43,20 @@
     coins_mined = 0
     # Run forever until interrupted
     while True:
-        # TODO: Get the last proof from the server and look for a new one
-        # TODO: When found, POST it to the server {"proof": new_proof}
-        # TODO: If the server responds with 'New Block Forged'
-        # add 1 to the number of coins mined and print it.  Otherwise,
-        # print the message from the server.
-        pass
+        last_proof = fetch_last_proof()
+        print("Calculating proof")
+        start_time = time.time()
+        proof = proof_of_work(last_proof)
+        end_time = time.time()
+        time_elapsed = end_time - start_time
+        print(f"Proof calculated in {time_elapsed} seconds.")
+        response = mine_block(proof)
+        if response:
```

Comment:
You did a good job solving this. Another option would have been to put everything into your while loop instead of calling functions. 

r = requests.get(url=node + "/last_proof")
data = r.json()
new_proof = proof_of_work(data.get('proof'))

post_data = {"proof": new_proof}

r = requests.post(url=node + "/mine", json=post_data)
data = r.json()
if data.get('message') == 'New Block Forged':
____coins_mined += 1
____print("Total coins mined: " + str(coins_mined))
else:
____print(data.get('message'))